File: Aula_11/src/android/app/src/main/java/com/example/locker/domain/LockError.kt
package com.example.locker.domain

import com.example.locker.protocol.LockProtocol

/**
 * Erros que o app sabe explicar ao usuário.
 *
 * Espelha `ios/SmartLock/Domain/LockError.swift`; a `message` é o texto exibido
 * na interface, então precisa fazer sentido para quem está diante da porta.
 */
sealed class LockError(message: String) : Exception(message) {
    /** Rádio desligado, sem permissão, ou aparelho sem BLE. */
    class BluetoothUnavailable(reason: String) : LockError(reason)

    object NotConnected : LockError("Sem conexão com a fechadura.")

    class CharacteristicMissing(name: String) :
        LockError("A fechadura não expõe a característica $name.")

    object Timeout : LockError("A fechadura não respondeu a tempo.")

    object Cancelled : LockError("Operação cancelada.")

    class UnsupportedProtocol(version: Int) : LockError(
        "Fechadura usa a versão $version do protocolo; o app usa a ${LockProtocol.VERSION}."
    )

    object MalformedResponse : LockError("Resposta da fechadura em formato inválido.")

    object NoCredential :
        LockError("Este aparelho ainda não tem credencial para esta fechadura.")

    class AccessDenied(reason: String?) :
        LockError(reason ?: "Acesso negado pela fechadura.")

    object RateLimited :
        LockError("Muitas tentativas inválidas. Aguarde antes de tentar de novo.")

    /** Falha no Keystore/armazenamento local da credencial. */
    class Storage(reason: String) : LockError("Falha ao guardar a credencial: $reason")

    /** Qualquer falha do canal BLE que não se encaixe nos casos acima. */
    class Transport(message: String) : LockError(message)
}

/** Texto pronto para a interface, mesmo para exceções que não são [LockError]. */
fun Throwable.userMessage(): String =
    message?.takeIf { it.isNotBlank() } ?: "Falha inesperada (${this::class.simpleName})."

File: Aula_11/src/ios/SmartLock/Domain/LockError.swift
import Foundation

enum LockError: LocalizedError, Equatable {
    case bluetoothUnavailable(String)
    case notConnected
    case characteristicMissing(String)
    case timeout
    case cancelled
    case unsupportedProtocol(Int)
    case malformedResponse
    case noCredential
    case accessDenied(String?)
    case rateLimited
    case keychain(OSStatus)
    case transport(String)

    var errorDescription: String? {
        switch self {
        case .bluetoothUnavailable(let reason):
            return reason
        case .notConnected:
            return "Sem conexão com a fechadura."
        case .characteristicMissing(let name):
            return "A fechadura não expõe a característica \(name)."
        case .timeout:
            return "A fechadura não respondeu a tempo."
        case .cancelled:
            return "Operação cancelada."
        case .unsupportedProtocol(let version):
            return "Fechadura usa a versão \(version) do protocolo; o app usa a \(LockProtocol.version)."
        case .malformedResponse:
            return "Resposta da fechadura em formato inválido."
        case .noCredential:
            return "Este iPhone ainda não tem credencial para esta fechadura."
        case .accessDenied(let reason):
            return reason ?? "Acesso negado pela fechadura."
        case .rateLimited:
            return "Muitas tentativas inválidas. Aguarde antes de tentar de novo."
        case .keychain(let status):
            return "Falha no Keychain (código \(status))."
        case .transport(let message):
            return message
        }
    }
}
